### DISCUSSION TITLE
Some questions regarding middleware

### DISCUSSION THREAD

**Discussion by imageck** [Automattic/mongoose#13082] [Q&A]
Howdies!

I never knew working with a database could be so fun! Mongoose is such a breath of fresh air.

Anyway, I was reading through the documentation and the concept of middleware has sparked my interest. Even though I haven't actually tried them yet I can see their use. Two questions remain:

1. Could someone please explain [this](https://mongoosejs.com/docs/middleware.html#defining) bit about exports to me?
> As a consequence, be careful about exporting Mongoose models from the same file that you define your schema. If you choose to use this pattern, you must define [global plugins](https://mongoosejs.com/docs/api/mongoose.html#mongoose_Mongoose-plugin) before calling require() on your model file.
> ```javascript
> const schema = new mongoose.Schema({ name: String });
> 
> // Once you `require()` this file, you can no longer add any middleware
> // to this schema.
> module.exports = mongoose.model('User', schema);
> ```
Does it simply imply that the set of middleware is immutable once the schema they're defined in is compiled the very first time? Meaning I won't be able to add new hooks or drop the existing ones should I decide to later on?

2. I have a collection of items with a `tags` key:
```javascript
  tags: {
    type: [String],
    validate: {
      validator: a => a.length <= 10
    }
  },
```
but I'd also like to able to store them in their own collection. So a post-save hook seems like a perfect fit here. What would be a good way of achieving that? Export the `Tag` model in `Item` and do something like `Tag.create()` in the post-save hook?

Thanks!

**Comment by vkarpov15** [comment]
Thanks for your kind words, we're flattered :)

1. In general Mongoose doesn't support modifying the model's schema after compiling the model using `mongoose.model()` or `Connection.prototype.model()`. So that's why we recommend being careful about exporting `mongoose.model()` calls. However, there's a lot of workarounds.
2. Have a separate `Tag` model. I think `Tag.findOneAndUpdate()` with upsert in the `Item` schema's post('save') hook would be the easiest way to keep the `Tag` model in sync with the `Item` model.

**Reply by imageck** [reply]
Perfect! Thanks for the tips!
